samples/benchmarks/coremark/src/main.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/logging/log_ctrl.h>
+#include <zephyr/drivers/gpio.h>
+
+#include "coremark_zephyr.h"
+
+LOG_MODULE_REGISTER(app, LOG_LEVEL_INF);
+
+#ifndef CONFIG_APP_MODE_FLASH_AND_RUN
+/*
+ * Get button configuration from the devicetree. This is mandatory.
+ */
+#define BUTTON_NODE             DT_ALIAS(button)
+#define LED_NODE                DT_ALIAS(led)
+
+#if !DT_NODE_HAS_STATUS(BUTTON_NODE, okay)
+	#error "Unsupported board: /"button/" alias is not defined."
+#endif
+
+#if !DT_NODE_HAS_STATUS(LED_NODE, okay)
+	#error "Unsupported board: /"led/" alias is not defined."
+#endif
+
+#define BUTTON_LABEL            DT_PROP(DT_ALIAS(button), label)
+#define LED_ON()                (void)gpio_pin_set_dt(&status_led, 1)
+#define LED_OFF()               (void)gpio_pin_set_dt(&status_led, 0)
+
+static const struct gpio_dt_spec start_button = GPIO_DT_SPEC_GET(BUTTON_NODE, gpios);
+static const struct gpio_dt_spec status_led = GPIO_DT_SPEC_GET(LED_NODE, gpios);
+
+#else
+
+#define BUTTON_LABEL            "Reset"
+#define LED_ON()
+#define LED_OFF()
+
+#endif
+
+static K_SEM_DEFINE(start_coremark, 0, 1);
+static atomic_t coremark_in_progress;
+
+static void flush_log(void)
+{
+	if (IS_ENABLED(CONFIG_LOG_PROCESS_THREAD)) {
+		while (log_data_pending()) {
+			k_sleep(K_MSEC(10));
+		}
+		k_sleep(K_MSEC(10));
+	} else {
+		while (LOG_PROCESS()) {
+		}
+	}
+}
+
+#ifndef CONFIG_APP_MODE_FLASH_AND_RUN
+
+static void button_pressed(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	if (gpio_pin_get_dt(&start_button) && !atomic_get(&coremark_in_progress)) {
+		LOG_INF("%s pressed!", BUTTON_LABEL);
+		(void)atomic_set(&coremark_in_progress, true);
```

Comment:
Actually for atomics we should use `atomic_cas` (get + set is a bad coding practice because it's no longer atomic operation)